pkg/monitoring/alloy/configmap.go

```diff
@@ -1,21 +1,182 @@
+		var monitoringConfig MonitoringConfig
+		err := yaml.Unmarshal([]byte(currentState.Data["values"]), &monitoringConfig)
+		if err != nil {
+			logger.Info("alloy-service - failed to unmarshal current monitoring config", "error", err)
+		} else {
+			currentShards = monitoringConfig.Alloy.Controller.Replicas
+			logger.Info("alloy-service - current number of shards", "shards", currentShards)
+		}
+	}
+
+	// Compute the number of shards based on the number of series.
+	query := fmt.Sprintf(`sum(max_over_time(prometheus_remote_write_wal_storage_active_series{cluster_id="%s", component_id="prometheus.remote_write.default", service="%s"}[6h]))`, cluster.Name, commonmonitoring.AlloyMonitoringAgentAppName)
+	headSeries, err := querier.QueryTSDBHeadSeries(ctx, query)
+	if err != nil {
+		logger.Error(err, "alloy-service - failed to query head series")
+		metrics.MimirQueryErrors.WithLabelValues().Inc()
+	}
+
+	clusterShardingStrategy, err := commonmonitoring.GetClusterShardingStrategy(cluster)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	shardingStrategy := a.MonitoringConfig.DefaultShardingStrategy.Merge(clusterShardingStrategy)
+	shards := shardingStrategy.ComputeShards(currentShards, headSeries)
+
+	alloyConfig, err := a.generateAlloyConfig(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	data := struct {
+		AlloyConfig       string
+		PriorityClassName string
+		Replicas          int
+		RequestsCPU       string
+		RequestsMemory    string
+		SecretName        string
+	}{
+		AlloyConfig:       alloyConfig,
+		PriorityClassName: commonmonitoring.PriorityClassName,
+		Replicas:          shards,
+		RequestsCPU:       commonmonitoring.AlloyRequestsCPU,
+		RequestsMemory:    commonmonitoring.AlloyRequestsMemory,
+		SecretName:        commonmonitoring.AlloyMonitoringAgentAppName,
+	}
+
+	var values bytes.Buffer
+	err = alloyMonitoringConfigTemplate.Execute(&values, data)
+	if err != nil {
+		return nil, err
+	}
+
+	configMapData := make(map[string]string)
+	configMapData["values"] = values.String()
+
+	return configMapData, nil
+}
+
+func (a *Service) generateAlloyConfig(ctx context.Context, cluster *clusterv1.Cluster) (string, error) {
+	var values bytes.Buffer
+
+	organization, err := a.OrganizationRepository.Read(ctx, cluster)
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+
+	provider, err := common.GetClusterProvider(cluster)
+	if err != nil {
+		return "", errors.WithStack(err)
+	}
+
+	data := struct {
+		RemoteWriteURLEnvVarName               string
+		RemoteWriteNameEnvVarName              string
+		RemoteWriteBasicAuthUsernameEnvVarName string
+		RemoteWriteBasicAuthPasswordEnvVarName string
+		RemoteWriteTimeout                     string
+		RemoteWriteTLSInsecureSkipVerify       bool
+
+		QueueConfigCapacity          int
+		QueueConfigMaxSamplesPerSend int
+		QueueConfigMaxShards         int
+
+		ExternalLabelsClusterID       string
```

Comment:
Could externallabels not be a map?

Reply:
Yes good point, I though about this at some point and got carried on with other stuff

Reply:
Maybe the scruct definition should be a named structure?

Reply:
I don't see a need for a named struct while we can use a map[string]string

Reply:
I meant for the parent struct here but sure https://github.com/giantswarm/observability-operator/pull/80/files/ec3ad04b8dd6308937f06db614ccb65c62b7d0dd#diff-1fd69b0b640bb9c306edbeee160e6dc6d5876d4bb36107ef49c13bad498e78baR79

Reply:
I have no strong opinion here.